File: json-schema/research/format_duration/runner.go
// Go runner for the `duration` format conformance corpus.
//
// Compiles the SINGLE generator-owned pinned regex (from corpus.json's
// `pinned_regex`) with Go's regexp (RE2) and, for each corpus value, reports
// whether it matches. The pinned regex is fully anchored (^...$), so a plain
// MatchString gives the whole-string verdict.
//
// Reads corpus.json (argv[1] or ./corpus.json) and emits JSON Lines to stdout:
//
//	{"id","engine":"go","compiled":bool,"matched":bool|null}
//
// `compiled` reports whether the pinned regex compiled at all (same value for
// every row) and `matched` is null only if it did not.
//
// Run: go run runner.go [corpus.json]
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
)

type kase struct {
	ID    string `json:"id"`
	Value string `json:"value"`
}

type corpus struct {
	PinnedRegex string `json:"pinned_regex"`
	Cases       []kase `json:"cases"`
}

type result struct {
	ID       string `json:"id"`
	Engine   string `json:"engine"`
	Compiled bool   `json:"compiled"`
	Matched  *bool  `json:"matched"`
}

func main() {
	path := "corpus.json"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	data, err := os.ReadFile(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, "read error:", err)
		os.Exit(1)
	}
	var c corpus
	if err := json.Unmarshal(data, &c); err != nil {
		fmt.Fprintln(os.Stderr, "parse error:", err)
		os.Exit(1)
	}
	re, compileErr := regexp.Compile(c.PinnedRegex)
	enc := json.NewEncoder(os.Stdout)
	for _, k := range c.Cases {
		r := result{ID: k.ID, Engine: "go"}
		if compileErr == nil {
			r.Compiled = true
			m := re.MatchString(k.Value)
			r.Matched = &m
		}
		_ = enc.Encode(r)
	}
}

File: json-schema/research/format_duration/runner.mjs
// JS runner (node) for the `duration` format conformance corpus.
//
// Compiles the SINGLE generator-owned pinned regex (from corpus.json's
// `pinned_regex`) with `new RegExp(p, "u")` (the u flag is the pinned choice
// for the TS/JS target) and tests each corpus value. The pinned regex is fully
// anchored (^...$), so `.test` gives the whole-string verdict.
//
// Reads corpus.json (argv[2] or ./corpus.json) and emits JSON Lines to stdout:
//   {"id","engine":"js","compiled":bool,"matched":bool|null}
//
// Run: node runner.mjs [corpus.json]
import { readFileSync } from "node:fs";

const path = process.argv[2] ?? "corpus.json";
const corpus = JSON.parse(readFileSync(path, "utf8"));

let re = null;
let compiled = false;
try {
  re = new RegExp(corpus.pinned_regex, "u");
  compiled = true;
} catch {
  compiled = false;
}

for (const k of corpus.cases) {
  const matched = compiled ? re.test(k.value) : null;
  process.stdout.write(
    JSON.stringify({ id: k.id, engine: "js", compiled, matched }) + "\n",
  );
}
